### ISSUE TITLE
Investigate ways to propagate the mTLS client identity through to the upstream cluster

### ISSUE THREAD

**Issue by k-wall** [kroxylicious/kroxylicious#1637]
With plain (unproxied) Kafka, users have the option to use TLS client authentication to authenticate kafka applications.  The client certificate's identity is made available within the Broker where it can be used for authorization decisions (ACLs) or audit purposes to understand what actions the user has performed.

If the user is proxying Kafka using Kroxylicious, they can opt to use TLS client auth (#1631) but there is currently no way to    propagate the client's identity through to the upstream cluster.

This task is to look at ways that it might be achieved.

**Comment by k-wall** [comment]
One idea:  utilise cooperating SASL/listener configuration on the upstream side to accept the client identity in a secure fashion.   You might be able to do this with a custom SASL mechanism  (to propagate the identity) coupled with a dedicated listener on the upstream clusters that authenticates the proxy using mTLS.

**Comment by callaertanthony** [comment]
From my understanding, problem to solve is not only on mTLS side. Currently Korxylicious doesn't manage `Principal` extraction mechanism even for SASL.
Maybe we must open the discussion on how we would like to manage `Principal` at the proxy level, and maybe add a logic with filters for forward and, why not, RBAC before upstream

**Comment by k-wall** [comment]
We have https://github.com/kroxylicious/kroxylicious/issues/1132 talking about giving the filters access to the Principal, but you are correct -  there is a Principal coming from SASL and a separate Principal coming from downstream TLS client auth.

I don't think there's an issue talking about _authorization_ offload but is certainly sounds like it would have some use-cases.

**Comment by ddaviesbrackett-dwp** [comment]
that new, cooperating authoriser would require some explicit (and preferably cryptographically strong) trust between the proxy and the brokers - to use the HTTP analogy, the `X-Forwarded-For` header is not a basis on which to make authorisation decisions. Not that the brokers don't already have to trust the proxy and its filters implicitly already - but if the broker listener that used the new cooperating authoriser were to (accidentally) accept connections from elsewhere than the proxy, that's a big authorisation hole.

**Comment by k-wall** [comment]
> Not that the brokers don't already have to trust the proxy and its filters implicitly already - but if the broker listener that used the new cooperating authoriser were to (accidentally) accept connections from elsewhere than the proxy, that's a big authorisation hole.

Yes, absolutely.    I was wondering about a pattern where TLS client auth is used between **proxy/broker** too.   The broker would have a listener configured to **require** TLS client auth and accept only connections presenting a certain cert - one issued exclusively to the proxy). The listener would be configured with a custom SASL mechanism accepted the principal relayed by it unconditionally.   The proxy would relay the principal from the downstream to the upstream via this SASL mechanism.

The disadvantage - that is a lot of configuration levers that needs to be gotten just right... otherwise you have a gaping security hole.

**Comment by ddaviesbrackett-dwp** [comment]
I follow your logic - that's what I was assuming, that there'd be a listener exclusively for the proxy's use configured to use the new authoriser. But it is a weighty footgun, one that almost makes me think that changing the auth requirements for clients to sasl_ssl (preferably with oauthbearer as the mechanism) might be less overall risk than adding this authoriser.

**Comment by SamBarker** [comment]
I'm quite sympathetic to the footgun argument, but I don't have mental model for an alternative.

>clients to sasl_ssl (preferably with oauthbearer as the mechanism) might be less overall risk than adding this authoriser.

@ddaviesbrackett-dwp can you expand a bit on how this works? Does it effectively de-couple connection auth from user auth? I.e. using m_tls to prove its a proxy communicating with the broker it expects to but that says nothing about the identity of the user making the connection?

**Comment by k-wall** [comment]
> I follow your logic - that's what I was assuming, that there'd be a listener exclusively for the proxy's use configured to use the new authoriser. But it is a weighty footgun, one that almost makes me think that changing the auth requirements for clients to sasl_ssl (preferably with oauthbearer as the mechanism) might be less overall risk than adding this authoriser.

I was making the assumption that TLS client authentication requirement between client and proxy was immutable.  

I have seen proxies in the HTTP domain (HaProxy) relay the client's CN to the upstream by injecting HTTP headers (`X-SSL-Client-CN`).   The idea above is basically this same idea re-expressed using the SASL Auth RPC as the relay mechanism, rather than a header.

**Comment by k-wall** [comment]
> using m_tls to prove its a proxy communicating with the broker it expects to but that says nothing about the identity of the user making the connection?

That's my mental model.   If my reading org.apache.kafka.common.network.ChannelBuilders#create  is correct, Kafka will prefer taking the Principal from SASL over TLS client auth.     So you could use TLS client auth to enforce that a TLS server talks only to TLS client(s) it trust and use SASL to establish identity on that channel.

**Comment by ddaviesbrackett-dwp** [comment]
Well, if HaProxy does it and that's trusted and accepted, then that's my argument shot! In our case, #1631 was more about belt-and-braces, requiring both authorisation-to-connect via client TLS certificates and authentication-and-authorisation-to-operate via SASL. 

> So you could use TLS client auth to enforce that a TLS server talks only to TLS client(s) it trusts and use SASL to establish identity on that channel.

much more succinctly put than I was able to manage, thanks Keith :)

**Comment by k-wall** [comment]
> In our case, https://github.com/kroxylicious/kroxylicious/pull/1631 was more about belt-and-braces, requiring both authorisation-to-connect via client TLS certificates and authentication-and-authorisation-to-operate via SASL.

actually, maybe that the best the best stance.  That avoids the security nightmares of the trusting listener on the broker etc

I'd love to hear what others think @kroxylicious/developers

**Comment by SamBarker** [comment]
>So you could use TLS client auth to enforce that a TLS server talks only to TLS client(s) it trusts and use SASL to establish identity on that channel.

Thats certainly a model that works for me. It fits with my (very hazy) mental model of what `m_tls` is used for in service mesh  deployments. It also avoids conflating two things: peer identity with user identity. 

So we are saying that: Kroxylicious should be able to do use `m_tls` with both upstream and downstream peers to validate that the connections are with the appropriate peers. However we only "support" using SASL_* for authentication & authorization in the broker.

That seems like a great starting point to me. 

We have some code in support of doing authentication offload. How would that fit with the above model? I wonder if "offload" is the wrong term / goal. What we really want is the proxy to ensure is that only clients with valid credentials make it through to the brokers, the broker can then authenticate and authorize as normal. Thus we remove the need for any co-operating listeners/plug-ins and the attendant security holes.
